saleor/graphql/discount/mutations/promotion/validators.py

```diff
@@ -57,6 +60,47 @@ def clean_promotion_rule(
     return cleaned_input
 
 
+def _get_gift_ids(cleaned_input, instance):
+    """Return the set of gift ids for promotion rule valid after performing mutation."""
+    if not instance and not any(
+        [field in cleaned_input for field in ["gifts", "add_gifts", "remove_gifts"]]
+    ):
+        return
+
+    if "gifts" in cleaned_input:
+        gift_ids = {gift.id for gift in cleaned_input["gifts"]}
+    else:
+        # this part is only for PromotionRuleUpdate mutation
+        # so the gifts will be fetched once
+        gift_ids = (
+            {gift.id for gift in instance.gifts.all()}
+            | {gift.id for gift in cleaned_input.get("add_gifts", [])}
+        ) - {gift.id for gift in cleaned_input.get("remove_gifts", [])}
+
+    return gift_ids
```

Comment:
This is complicated already, maybe we can get rid of the variable? What both cases do is return list of ids at 1st glance so we don't loose much by not assigning those to named variable.
```suggestion
    if "gifts" in cleaned_input:
        return {gift.id for gift in cleaned_input["gifts"]}
    else:
        # this part is only for PromotionRuleUpdate mutation
        # so the gifts will be fetched once
        return (
            {gift.id for gift in instance.gifts.all()}
            | {gift.id for gift in cleaned_input.get("add_gifts", [])}
        ) - {gift.id for gift in cleaned_input.get("remove_gifts", [])}
```

Reply:
I removed `gift_ids` variable and created a variable for each of sets to make the calculation more clear. I hope it helps